Do not use AbstractTag<T> for Span.setTag(), but the actual impls.

carlosalberto opened the pull request:
This is done to not use an abstract class for our public API,
so we take overloads of BooleanTag, IntTag and StringTag.

Addresses a small detail mentioned in #314 

@yurishkuro @tylerbenson

yurishkuro commented:
Why not use a typed interface? I'm not a fan of overloaded methods (even though it is consistent with the plain set methods).

carlosalberto commented:
@yurishkuro 

You refer to something like this?

```java
// IntTag and the others would be inheriting this class.
public class TypedTag<T> extends AbstractTag<T> {
    protected void set(Span span, T tagValue)
    {   
        span.setTag(super.key, tagValue.toString());
    }
}
```

carlosalberto commented:
Hey @yurishkuro sorry, re-read your message today and figured you simply wanted the same, but through an interface ;)

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/tag/Tag.java:
```diff
@@ -0,0 +1,20 @@
+import io.opentracing.Span;
+
+public interface Tag<T> {
+    String getKey();
```
How about setValue(Span, T) instead?

carlosalberto replied:
Had considered the same, but was afraid of making `AbstractTag.set()` public (right now is protected) ;)

Although now I think we could leave `set()` as abstract and have `setValue()`, as an alternative ;) Would that work for you?

yurishkuro replied:
I would rather make set() public. I don't know why it's protected now, since all concrete classes make it public.

yurishkuro commented on opentracing-mock/src/main/java/io/opentracing/mock/MockSpan.java:
```diff
@@ -137,8 +137,8 @@ public MockSpan setTag(String key, Number value) {
-    public <T> Span setTag(AbstractTag<T> key, T value) {
-        return setObjectTag(key.getKey(), value);
+    public <T> MockSpan setTag(Tag<T> tag, T value) {
+        return setObjectTag(tag.getKey(), value);
```
If you add setValue to the interface you won't need to use object.

carlosalberto commented:
Hey @yurishkuro 

I added `Tag.set()` and updated `AbstractTag` as well - but left the call to `MockTracer.setObjectTag()` as this extra checks that no tags are added after a `Span` was finished, and would love to keep it around:

```java
    // An error will be reported if this Span is already finished.
    private synchronized MockSpan setObjectTag(String key, Object value) {
        finishedCheck("Adding tag {%s:%s} to already finished span", key, value);
```

Let me know ;)

yurishkuro commented:
I think it should do `tag.set(span, value)` instead. This way the correct typed set() method will be called by the tag itself, and _those_ methods would call setObjectTag

carlosalberto commented:
@yurishkuro Updated it to looks just like that ;) If you think that's fine, I will merge soon

yurishkuro commented:
🚢

carlosalberto commented:
👍
